File: rust_cpp/md.cpp
// binance_book_standalone.cpp
#include <boost/asio.hpp>
#include <boost/asio/ssl/context.hpp>
#include <boost/beast/core.hpp>
#include <boost/beast/ssl.hpp>
#include <boost/beast/websocket.hpp>
#include <boost/beast/http.hpp>
#include <nlohmann/json.hpp>

#include <atomic>
#include <chrono>
#include <csignal>
#include <iomanip>
#include <iostream>
#include <map>
#include <mutex>
#include <sstream>
#include <string>
#include <thread>
#include <vector>
#include <algorithm>
#include <cmath>

using tcp = boost::asio::ip::tcp;
namespace ssl = boost::asio::ssl;
namespace http = boost::beast::http;
using json = nlohmann::json;

struct OrderBook {
  std::map<double,double,std::greater<double>> bids;
  std::map<double,double,std::less<double>> asks;
};

using Clock = std::chrono::steady_clock;
inline long long to_us(Clock::duration d) {
  return std::chrono::duration_cast<std::chrono::microseconds>(d).count();
}
inline long long now_ms() {
  return std::chrono::duration_cast<std::chrono::milliseconds>(
           std::chrono::system_clock::now().time_since_epoch()).count();
}

constexpr std::size_t UPDATE_LIMIT = 100;

template <typename Cmp>
void apply_deltas(std::map<double,double,Cmp>& side, const std::vector<std::pair<double,double>>& deltas) {
  for (auto& [p, sz] : deltas) {
    if (sz == 0.0) { auto it = side.find(p); if (it != side.end()) side.erase(it); }
    else side[p] = sz;
  }
}

std::atomic<bool> g_stop{false};
void handle_signal(int) { g_stop.store(true); }

std::string to_lower(std::string s) { for (auto& c : s) c = std::tolower(c); return s; }

bool fetch_snapshot(const std::string& symbol, OrderBook& out_book, long long& last_update_id, std::string& err) {
  try {
    boost::asio::io_context ioc;
    ssl::context ctx{ssl::context::tls_client};
    ctx.set_default_verify_paths();

    tcp::resolver resolver{ioc};
    boost::beast::ssl_stream<boost::beast::tcp_stream> stream{ioc, ctx};

    const std::string host = "api.binance.com";
    const std::string port = "443";
    const std::string target = "/api/v3/depth?symbol=" + symbol + "&limit=1000";

    auto const results = resolver.resolve(host, port);
    boost::beast::get_lowest_layer(stream).connect(results);
    if (!SSL_set_tlsext_host_name(stream.native_handle(), host.c_str())) throw std::runtime_error("SNI set failed");
    stream.handshake(ssl::stream_base::client);

    http::request<http::string_body> req{http::verb::get, target, 11};
    req.set(http::field::host, host);
    req.set(http::field::user_agent, "beast");
    http::write(stream, req);

    boost::beast::flat_buffer buffer;
    http::response<http::string_body> res;
    http::read(stream, buffer, res);
    boost::system::error_code ec;
    stream.shutdown(ec);

    if (res.result() != http::status::ok) { err = "HTTP " + std::to_string(static_cast<int>(res.result())); return false; }

    auto j = json::parse(res.body());
    last_update_id = j.at("lastUpdateId").get<long long>();
    out_book.bids.clear(); out_book.asks.clear();
    for (auto& lvl : j["bids"]) {
      double p = std::stod(lvl[0].get<std::string>()); double s = std::stod(lvl[1].get<std::string>());
      if (s > 0) out_book.bids[p] = s;
    }
    for (auto& lvl : j["asks"]) {
      double p = std::stod(lvl[0].get<std::string>()); double s = std::stod(lvl[1].get<std::string>());
      if (s > 0) out_book.asks[p] = s;
    }
    return true;
  } catch (const std::exception& e) { err = e.what(); return false; }
}

void apply_update_json(const std::string& payload, long long& last_update_id, OrderBook& book) {
  auto j = json::parse(payload);
  long long U = j.at("U").get<long long>();
  long long u = j.at("u").get<long long>();
  long long pu = j.contains("pu") ? j["pu"].get<long long>() : 0LL;

  if (u <= last_update_id) return;
  bool bridged_window = (U <= last_update_id + 1 && last_update_id + 1 <= u);
  bool bridged_pu = (pu == last_update_id && pu != 0);
  if (!(bridged_window || bridged_pu)) throw std::runtime_error("sequence gap; need resnapshot");

  std::vector<std::pair<double,double>> bid_d, ask_d;
  for (auto& lvl : j["b"]) { bid_d.emplace_back(std::stod(lvl[0].get<std::string>()), std::stod(lvl[1].get<std::string>())); }
  for (auto& lvl : j["a"]) { ask_d.emplace_back(std::stod(lvl[0].get<std::string>()), std::stod(lvl[1].get<std::string>())); }

  apply_deltas(book.bids, bid_d);
  apply_deltas(book.asks, ask_d);
  last_update_id = u;
}

void print_book(const OrderBook& book, std::size_t depth = 10, long long latency_us = -1, long long recv_ts_ms = -1) {
  auto _flags = std::cout.flags(); auto _prec = std::cout.precision();
  std::cout.setf(std::ios::fixed); std::cout << std::setprecision(6);

  std::cout << "[BINANCE]";
  if (recv_ts_ms >= 0) std::cout << " ts_ms=" << recv_ts_ms;
  if (latency_us >= 0) std::cout << " proc_us=" << latency_us;
  std::cout << " top " << depth << " levels\n  Bids: ";
  std::size_t i = 0;
  for (auto& [p, s] : book.bids) { if (i++ >= depth) break; std::cout << p << "@" << s << "  "; }
  if (i == 0) std::cout << "(empty)";
  std::cout << "\n  Asks: ";
  i = 0;
  for (auto& [p, s] : book.asks) { if (i++ >= depth) break; std::cout << p << "@" << s << "  "; }
  if (i == 0) std::cout << "(empty)";
  std::cout << "\n" << std::endl;

  std::cout.flags(_flags); std::cout.precision(_prec);
}

long long percentile_us(std::vector<long long>& sorted, double pct) {
  if (sorted.empty()) return 0;
  auto idx = static_cast<std::size_t>(std::llround((pct / 100.0) * (static_cast<double>(sorted.size() - 1))));
  return sorted[std::min(idx, sorted.size() - 1)];
}

void print_latency_stats(const std::vector<long long>& latencies) {
  if (latencies.empty()) {
    std::cout << "[STATS] no samples collected" << std::endl;
    return;
  }
  std::vector<long long> v = latencies;
  std::sort(v.begin(), v.end());
  long long min = v.front();
  long long max = v.back();
  long long p10 = percentile_us(v, 10.0);
  long long p50 = percentile_us(v, 50.0);
  long long p90 = percentile_us(v, 90.0);
  long long p99 = percentile_us(v, 99.0);
  std::cout << "[STATS] samples=" << v.size()
            << " min=" << min << "us"
            << " max=" << max << "us"
            << " p10=" << p10 << "us"
            << " p50=" << p50 << "us"
            << " p90=" << p90 << "us"
            << " p99=" << p99 << "us"
            << std::endl;
}

int main(int argc, char** argv) {
  std::string symbol = "BTCUSDT";
  if (argc > 1 && argv[1]) symbol = argv[1];

  std::signal(SIGINT, handle_signal);
  std::signal(SIGTERM, handle_signal);

  std::cout << "Starting Binance depth stream for " << symbol << " (Ctrl+C to quit)" << std::endl;

  std::vector<long long> latencies;
  latencies.reserve(UPDATE_LIMIT);

  using namespace std::chrono_literals;
  while (!g_stop.load()) {
    try {
      OrderBook book;
      long long last_id = 0;

      // 1) Connect WS first
      boost::asio::io_context ioc;
      ssl::context ctx{ssl::context::tls_client};
      ctx.set_default_verify_paths();
      tcp::resolver resolver{ioc};
      boost::beast::websocket::stream<boost::beast::ssl_stream<boost::beast::tcp_stream>> ws{ioc, ctx};

      const std::string host = "stream.binance.com";
      const std::string port = "9443";
      const std::string path = "/ws/" + to_lower(symbol) + "@depth@100ms";

      auto const results = resolver.resolve(host, port);
      boost::beast::get_lowest_layer(ws).connect(results);
      if (!SSL_set_tlsext_host_name(ws.next_layer().native_handle(), host.c_str())) throw std::runtime_error("SNI set failed");
      ws.next_layer().handshake(ssl::stream_base::client);
      ws.handshake(host, path);

      // 2) REST snapshot
      std::string err;
      if (!fetch_snapshot(symbol, book, last_id, err)) {
        std::cerr << "[BINANCE] snapshot error: " << err << std::endl;
        boost::system::error_code ec;
        ws.close(boost::beast::websocket::close_code::normal, ec);
        std::this_thread::sleep_for(1s);
        continue;
      }

      // 3) Bridge buffered updates
      boost::beast::flat_buffer buffer;
      for (;;) {
        buffer.clear();
        ws.read(buffer);
        auto t_recv = Clock::now();
        auto recv_ms = now_ms();
        std::string payload = boost::beast::buffers_to_string(buffer.data());
        try {
          auto before = last_id;
          apply_update_json(payload, last_id, book);
          if (last_id != before) { 
            auto proc = to_us(Clock::now() - t_recv);
            latencies.push_back(proc);
            print_book(book, 10, proc, recv_ms);
            if (latencies.size() >= UPDATE_LIMIT) { 
              print_latency_stats(latencies); 
              g_stop.store(true);
              break;
            }
            break; 
          }
        } catch (...) { continue; }
      }

      if (g_stop.load()) break;

      // 4) Steady-state
      for (;;) {
        if (g_stop.load()) break;
        buffer.clear();
        ws.read(buffer);
        auto t_recv = Clock::now();
        auto recv_ms = now_ms();
        std::string payload = boost::beast::buffers_to_string(buffer.data());
        try {
          apply_update_json(payload, last_id, book);
          auto proc = to_us(Clock::now() - t_recv);
          latencies.push_back(proc);
          print_book(book, 10, proc, recv_ms);
          if (latencies.size() >= UPDATE_LIMIT) { 
            print_latency_stats(latencies); 
            g_stop.store(true);
            break; 
          }
        } catch (const std::exception& e) {
          std::cerr << "[BINANCE] " << e.what() << " — resyncing..." << std::endl;
          break;
        }
      }

      boost::system::error_code ec;
      ws.close(boost::beast::websocket::close_code::normal, ec);
    } catch (const std::exception& e) {
      if (g_stop.load()) break;
      std::cerr << "[BINANCE] exception: " << e.what() << std::endl;
    }
    std::this_thread::sleep_for(1s);
  }

  std::cout << "Stopped" << std::endl;
  return 0;
}

File: rust_cpp/md.rs
use std::{
    collections::BTreeMap,
    time::{Duration, Instant, SystemTime, UNIX_EPOCH},
};

use futures_util::StreamExt;
use serde::Deserialize;
use tokio::{signal, time::sleep};
use tokio_tungstenite::{connect_async, tungstenite::Message};

#[derive(Default, Debug, Clone)]
struct OrderBook {
    bids: BTreeMap<f64, f64>, // descending iteration via rev()
    asks: BTreeMap<f64, f64>, // ascending
}

fn apply_deltas(side: &mut BTreeMap<f64, f64>, deltas: &[(f64, f64)], reverse: bool) {
    for &(p, sz) in deltas {
        if sz == 0.0 {
            side.remove(&p);
        } else {
            side.insert(p, sz);
        }
    }
    if reverse {
        // keep as-is; we’ll iterate with .iter().rev()
    }
}

#[derive(Deserialize)]
struct Snapshot {
    lastUpdateId: i64,
    bids: Vec<[String; 2]>,
    asks: Vec<[String; 2]>,
}

#[derive(Deserialize)]
struct WsUpdate {
    U: i64,
    u: i64,
    #[serde(default)]
    pu: i64,
    b: Vec<[String; 2]>,
    a: Vec<[String; 2]>,
}

fn now_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis() as i64
}

fn parse_side(raw: &[[String; 2]]) -> Vec<(f64, f64)> {
    raw.iter()
        .map(|arr| (arr[0].parse().unwrap_or(0.0), arr[1].parse().unwrap_or(0.0)))
        .collect()
}

fn print_book(book: &OrderBook, depth: usize, proc_us: i64, recv_ts_ms: i64) {
    print!("[BINANCE] ts_ms={} proc_us={} top {} levels\n  Bids: ", recv_ts_ms, proc_us, depth);
    let mut i = 0;
    for (p, s) in book.bids.iter().rev() {
        if i >= depth { break; }
        print!("{:.6}@{:.6}  ", p, s);
        i += 1;
    }
    if i == 0 { print!("(empty)"); }
    print!("\n  Asks: ");
    i = 0;
    for (p, s) in book.asks.iter() {
        if i >= depth { break; }
        print!("{:.6}@{:.6}  ", p, s);
        i += 1;
    }
    if i == 0 { print!("(empty)"); }
    print!("\n\n");
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let symbol = std::env::args().nth(1).unwrap_or_else(|| "BTCUSDT".to_string());
    let ws_symbol = symbol.to_lowercase();

    let mut term = signal::ctrl_c();
    println!("Starting Binance depth stream for {} (Ctrl+C to quit)", symbol);

    while term.try_recv().is_err() {
        if let Err(e) = run_once(&symbol, &ws_symbol).await {
            eprintln!("[BINANCE] error: {e} — retrying...");
            sleep(Duration::from_secs(1)).await;
        }
    }
    println!("Stopped");
    Ok(())
}

async fn run_once(symbol: &str, ws_symbol: &str) -> anyhow::Result<()> {
    // 1) Connect WS first
    let url = format!("wss://stream.binance.com:9443/ws/{}@depth@100ms", ws_symbol);
    let (ws_stream, _) = connect_async(&url).await?;
    let (mut ws_write, mut ws_read) = ws_stream.split();

    // 2) REST snapshot
    let snap_url = format!("https://api.binance.com/api/v3/depth?symbol={}&limit=1000", symbol);
    let snap: Snapshot = reqwest::get(&snap_url).await?.json().await?;
    let mut book = OrderBook::default();
    for (p, s) in parse_side(&snap.bids) {
        if s > 0.0 { book.bids.insert(p, s); }
    }
    for (p, s) in parse_side(&snap.asks) {
        if s > 0.0 { book.asks.insert(p, s); }
    }
    let mut last_id = snap.lastUpdateId;

    // 3) Bridge buffered updates until we get one that advances last_id
    loop {
        let msg = ws_read.next().await.ok_or_else(|| anyhow::anyhow!("WS closed"))??;
        let t_recv = Instant::now();
        let recv_ms = now_ms();
        let update = match msg {
            Message::Text(txt) => serde_json::from_str::<WsUpdate>(&txt)?,
            Message::Binary(bin) => serde_json::from_slice::<WsUpdate>(&bin)?,
            _ => continue,
        };
        apply_update(&update, &mut last_id, &mut book)?;
        print_book(&book, 10, t_recv.elapsed().as_micros() as i64, recv_ms);
        break;
    }

    // 4) Steady-state
    while let Some(msg) = ws_read.next().await {
        let msg = msg?;
        let t_recv = Instant::now();
        let recv_ms = now_ms();
        match msg {
            Message::Text(txt) => {
                let update: WsUpdate = serde_json::from_str(&txt)?;
                apply_update(&update, &mut last_id, &mut book)?;
                print_book(&book, 10, t_recv.elapsed().as_micros() as i64, recv_ms);
            }
            Message::Binary(bin) => {
                let update: WsUpdate = serde_json::from_slice(&bin)?;
                apply_update(&update, &mut last_id, &mut book)?;
                print_book(&book, 10, t_recv.elapsed().as_micros() as i64, recv_ms);
            }
            Message::Ping(p) => ws_write.send(Message::Pong(p)).await?,
            Message::Close(_) => anyhow::bail!("WS closed"),
            _ => {}
        }
    }
    Ok(())
}

fn apply_update(update: &WsUpdate, last_id: &mut i64, book: &mut OrderBook) -> anyhow::Result<()> {
    let U = update.U;
    let u = update.u;
    let pu = update.pu;

    if u <= *last_id {
        return Ok(());
    }
    let bridged_window = U <= *last_id + 1 && *last_id + 1 <= u;
    let bridged_pu = pu != 0 && pu == *last_id;
    if !(bridged_window || bridged_pu) {
        anyhow::bail!("sequence gap; need resnapshot");
    }

    let bid_d = parse_side(&update.b);
    let ask_d = parse_side(&update.a);
    apply_deltas(&mut book.bids, &bid_d, true);
    apply_deltas(&mut book.asks, &ask_d, false);
    *last_id = u;
    Ok(())
}

